### PR TITLE
Move PerKeyCombineFnRunners to runners/core.

### PR DESCRIPTION
**Pull request by peihe**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by peihe** [comment]
R: @kennknowles 

This requires a dataflow worker update.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java [hunk 1]
```diff
@@ -2389,15 +2388,31 @@ private GroupedValues(
...
           new OldDoFn<KV<K, ? extends Iterable<InputT>>, KV<K, OutputT>>() {
             @Override
-            public void processElement(ProcessContext c) {
+            public void processElement(final ProcessContext c) {
               K key = c.element().getKey();
 
-              c.output(KV.of(key, combineFnRunner.apply(key, c.element().getValue(), c)));
+              OutputT output;
+              if (fn instanceof RequiresContextInternal) {
+                output = ((KeyedCombineFnWithContext<? super K, ? super InputT, ?, OutputT>) fn)
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java]: Even though it is nice to have semantic marker interfaces like `RequiresContextInternal`, I think that the `instanceof` check and the cast should match.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java [hunk 2]
```diff
@@ -2389,15 +2388,31 @@ private GroupedValues(
...
+                        return c.getPipelineOptions();
+                      }
+
+                      @Override
+                      public <T> T sideInput(PCollectionView<T> view) {
+                        return c.sideInput(view);
+                      }
+                    });
+              } else {
+                output = ((KeyedCombineFn<? super K, ? super InputT, ?, OutputT>) fn)
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java]: I would actually also do `else if(fn instanceof KeyedCombineFn)` here prior to cast, and then add an `else` branch with a more informative error message like "got a `CombineFn` that I don't know how to handle".

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java (hunk 1 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java (hunk 2 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java]: done

**Comment by kennknowles** [comment]
LGTM. I will merge after confirming integration testing.

**Comment by peihe** [comment]
PTAL
Tests passed.

**Pull request closed without merging**